### PR TITLE
Improve errors for misused attributes

### LINKED ISSUE #105: Error when attributes have been misused
Find and fix all the [`Report this isn't valid here` comments](https://github.com/shepmaster/snafu/blob/52e16704228abb9310f74ba74f1b755383c17ada/snafu-derive/src/lib.rs#L165). Super annoying!

### PR DESCRIPTION
**Pull request by tjkirch**
Fixes #105.

The first commit directly addresses the discussion in #105 by adding errors for cases of attribute misuse that were known to be wrong but didn't yet have errors.  Previously, the derive would just continue without the effect the user was trying to request.

The second commit addresses similar comments elsewhere in the file that suggested we warn if attributes are duplicated.  This wasn't discussed in #105 but seemed closely related.

---

**Testing done:**

Existing unit/doc tests still pass, and I added compile-fail tests to exercise each of the new error cases.

### CODE REVIEW

**Comment by shepmaster** [comment]
> The invalid cases can't all be tested together, because the struct derive depends on the enum being valid

Yep, there’s “phases” of compilation where we can’t continue. My hope is basically to make those phases as few and large as possible, but there will always be *some* case where it just doesn’t work.

> * throwing warnings, but it seemed better to me to throw errors

I don’t know that there’s an actual way to generate warnings, either…

> * Should there be unit tests showing that these new cases throw errors

I’m sorry that I forgot to mention this! Take a look at the [compile-fail tests](https://github.com/shepmaster/snafu/tree/master/compatibility-tests/compile-fail/tests/ui). These assert that the error messages continue to work, and I strongly agree that we should have tests for each of the new errors.

I don’t know yet if “more errors in one file” or “more files” is a better approach. Probably more files, but I have a feeling that will be slower in CI at some point. For now, I’d suggest creating at least one new file for these errors.

You’ll want to rebase on master to get the fixes to these tests (causing your CI failure)

> * are there any concerns about existing use cases breaking because we throw new errors

This is where warnings would be nice. They could be warnings for the current semver release, and made into errors for the next breaking change. 

Since we are still pre-1.0, I don’t mind bumping the version early and often, so I’d probably move to 0.5 after this is merged if we can’t figure out warnings. If it was already 1.0, I’d be more torn about a major bump for such a thing, as it was code that wasn’t working the way people intended anyway.

**Comment by tjkirch** [comment]
> compile-fail tests

Ah, thanks!  Hopefully I can add these tonight.

> probably move to 0.5 after this is merged

Would it help if I add commits for the update?  It looks like it'd be `snafu-derive/Cargo.toml`, `Cargo.toml`, `CHANGELOG.md`, and `src/guide/upgrading.md`.  I'm not sure what else you'd like in `CHANGELOG` but I could start it...

Or, that could wait until we have more research on warning capabilities.

**Comment by shepmaster** [comment]
> if I add commits for the update

Thank you, but that’s all right! I’m lazy and don’t keep the changelog up-to-date with each PR, only doing it when I do a release. It does provide a good opportunity to review the changes and make sure that nothing obvious is missing or untested.

> `CHANGELOG.md` and `src/guide/upgrading.md`

Feel free to suggest some text that you’d like to see as a seed for what will end up here!

**Comment by tjkirch** [comment]
Thanks for all your guidance @shepmaster!  I marked this as ready for review, since I think I've addressed your concerns, at least for an initial implementation.

File: compatibility-tests/compile-fail/tests/ui/attribute-duplication.rs [hunk 1]
```diff
@@ -0,0 +1,28 @@
+mod enum_duplication {
```
Comment (shepmaster) [inline compatibility-tests/compile-fail/tests/ui/attribute-duplication.rs]: Did you also want a module for structs? Otherwise, I think it’s ok to remove the module here.

File: snafu-derive/src/lib.rs [hunk 2]
```diff
@@ -165,8 +167,18 @@ fn parse_snafu_enum(
                 match attr {
                     SnafuAttribute::Display(d) => display_format = Some(d),
                     SnafuAttribute::Visibility(v) => visibility = Some(v),
-                    SnafuAttribute::Source(..) => { /* Report this isn't valid here? */ }
-                    SnafuAttribute::Backtrace(..) => { /* Report this isn't valid here? */ }
+                    SnafuAttribute::Source(..) => {
+                        errors.push(syn::Error::new(
+                            name.span(),
+                            "'source' attribute is only valid on individual fields; attempted to place on this variant:",
+                        ));
```
Comment (shepmaster) [inline snafu-derive/src/lib.rs]: I wonder if there’s anything we can do about the “programmer overhead” of creating these, as well as the duplicated phrases. Potentially, this could look like:

```rust
errors.only_on_field(name, “source”)
errors.only_on_field(name, “backtrace”)
```

As a minimum, this could be used to hide the inner `Vec` and construction of `Error`:

```rust
impl MyErrorCollectionName {
    fn add(&mut self, span: impl Spanned, text: impl Into<String>) { /* … */ }

    fn finish(self) -> Result<(), Vec<…>> { /* … */ }
}
```

File: snafu-derive/src/lib.rs [hunk 3]
```diff
@@ -262,15 +284,46 @@ fn parse_snafu_enum(
                 }
             }
 
+            // Reverse the lists of fields where we may find duplicates so that we can call the
+            // second (and later) items the "duplicates" without uglier indexing.
```
Comment (shepmaster) [inline snafu-derive/src/lib.rs]: I like it 😁

File: snafu-derive/src/lib.rs [hunk 4]
```diff
@@ -262,15 +284,46 @@ fn parse_snafu_enum(
...
-            // Report a warning if there are multiple?
-            // Report a warning if delegating and our own?
+            for extra_field in backtrace_delegates.into_iter() {
+                errors.push(syn::Error::new(
+                    extra_field.name.span(),
+                    "Only one field can be designated as 'backtrace(delegate)'; found a duplicate here:",
+                ));
+            }
+
+            if backtrace_field.is_some() && backtrace_delegate.is_some() {
```
Comment (shepmaster) [inline snafu-derive/src/lib.rs]: Can this be something like `if let Some(backtrace_field,  backtrace_delegate) = (backtrace_field.as_ref(), backtrace_delegate.as_ref())` to avoid the `unwrap`s?

File: snafu-derive/src/lib.rs [hunk 5]
```diff
@@ -305,19 +363,41 @@ fn parse_snafu_struct(
...
+                ));
+            }
             SnafuAttribute::Source(ss) => {
                 for s in ss {
                     match s {
-                        Source::Flag(..) => { /* Report this isn't valid here? */ }
+                        Source::Flag(..) => {
+                            errors.push(syn::Error::new(
+                                name.span(),
+                                "'source(bool)' is only valid on individual fields; attempted to place on this struct:",
```
Comment (shepmaster) [inline snafu-derive/src/lib.rs]: This one sticks out because it’s the only one that says `(bool)` — why the difference?

**Comment by tjkirch** [comment]
> There’s one place where I was lazy and didn’t put the “report this” text:
> 
> https://github.com/shepmaster/snafu/blob/da2d4aced258bdc443c90d8b3de6b127912e2af0/snafu-derive/src/lib.rs#L147-L151
> 
> Can you expand that code to also report the attributes?

To make sure I understand this - you mean changing from this functional style to a loop that checks all of the attributes from `attributes_from_syn` and adds errors for anything that's not `visibility`?

File: compatibility-tests/compile-fail/tests/ui/attribute-duplication.rs (hunk 1 above)
Comment (tjkirch) [reply compatibility-tests/compile-fail/tests/ui/attribute-duplication.rs]: I think there should be a module for structs, but I didn't get struct duplication checks working - the code structure for structs wasn't laid out the way it was for enums, and I was having a harder time figuring out how to do that.  I wanted to leave this module so it'd be clearer that enums aren't the only things to check.

**Comment by shepmaster** [comment]
> changing from this functional style to a loop

Yep. I don’t see an obvious way, but if you see something that’s a bit more functional-style, feel free to go for it!

File: snafu-derive/src/lib.rs (hunk 2 above)
Comment (tjkirch) [reply snafu-derive/src/lib.rs]: I lean toward the latter because I think having a custom message can be helpful for clarifying what the user actually did that wasn't quite right.  I'll give that a shot.

File: snafu-derive/src/lib.rs (hunk 4 above)
Comment (tjkirch) [reply snafu-derive/src/lib.rs]: Ah, I didn't know that syntax would work for combining `Option`s - I'll give it a shot.

File: snafu-derive/src/lib.rs (hunk 5 above)
Comment (tjkirch) [reply snafu-derive/src/lib.rs]: This one was confusing to me.  I was trying to figure out what `Flag` is, and I think it's when you use `source(false)` to say a field isn't the source even though it's named `source`.  You'd just say `source` if it *is* the source, so `source(true)` shouldn't really be necessary.  That's my understanding of this code path, and I used `source(bool)` as a shorthand for it, since I believe it would only happen if the user did specify a boolean inside `source()`.  Is that right?

File: snafu-derive/src/lib.rs (hunk 2 above)
Comment (shepmaster) [reply snafu-derive/src/lib.rs]: > I lean toward the latter

I don’t think they have to be either-or:

```rust
impl MyErrorCollectionName {
    fn add_only_on_field(&mut self, span: impl Spanned, field: &str) {
        self.add(span, format!(“This is repeated text for field {}, do this other thing instead”));
    }

    fn add(&mut self, span: impl Spanned, text: impl Into<String>) { /* … */ }
}
```

Comment (tjkirch) [reply snafu-derive/src/lib.rs]: Ahh, I see what you mean.  "field" works well for handling duplication, but we might also want something like "element type" for misuse?  I think I'd typically want to say something like "this doesn't apply to [field, variant, enum, struct], it's only valid on [field, variant, enum, struct]".  I guess that could just be handled by the custom message, but it's even more common than repetition, so maybe it's worth a try simplifying it too.

Comment (shepmaster) [reply snafu-derive/src/lib.rs]: Yeah, there's lots of ways this could go. My suggestion would be to start with adding the plain `add` method and then see what repetition remains. That might then highlight new methods to be created. How those look could be as simple as taking in a `&str`, up to taking in a slice of items, or as far as some custom type(s). I think it's worth staying practical with respect to the amount of code we are saving, so the latter option seems unlikely :-) 

As a side note, you can't actually use the `impl ...` syntax I listed because we support Rust 1.18, but hopefully the idea is communicated.

File: snafu-derive/src/lib.rs (hunk 5 above)
Comment (shepmaster) [reply snafu-derive/src/lib.rs]: Ah, yes, I see. Your analysis is correct. I was missing that `source(from(...))` was valid here, but `source(false)` wasn't.

> so `source(true)` shouldn't really be necessary

I did this for consistency of `true` / `false` but also *potentially* allowing the concept of multiple attributes that subsequently override each other. There's also a nascent idea of allowing multiple `source(from())` attributes, for example. I wouldn't worry about either of those future plans, but just so you are aware of the background.

File: snafu-derive/src/lib.rs (hunk 2 above)
Comment (tjkirch) [reply snafu-derive/src/lib.rs]: I started with `add`, moved to a couple custom methods that covered 90% of the cases, and eventually found that a couple custom types were simpler.  While it's slightly more verbose, I think it's rather clear - hope you like it as much as I do :)

File: snafu-derive/src/lib.rs [hunk 6]
```diff
@@ -98,6 +99,76 @@ impl Transformation {
...
+    /// The name of the attribute that was misused.
+    attribute: &'static str,
+    /// A description of where that attribute is valid.
+    valid_on: &'static str,
+    /// A description of where the attribute was incorrectly used.
+    not_on: &'static str,
+}
+
+impl fmt::Display for OnlyValidOn {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
```
Comment (shepmaster) [inline snafu-derive/src/lib.rs]: Sadly, `Foo<'_>` isn't valid in older versions of Rust, so we'll have to leave that off for now.

File: snafu-derive/src/lib.rs [hunk 7]
```diff
@@ -98,6 +99,76 @@ impl Transformation {
...
 
+/// SyntaxErrors is a convenience wrapper for a list of syntax errors discovered while parsing
+/// something that derives Snafu.  It makes it easier for developers to add and return syntax
+/// errors while walking through the parse tree.
+struct SyntaxErrors {
+    inner: Vec<syn::Error>,
+}
+
+impl SyntaxErrors {
+    fn new() -> SyntaxErrors {
```
Comment (shepmaster) [inline snafu-derive/src/lib.rs]: This is fine, but for what it's worth, I would have just added `#[derive(Default)]` and then called `SyntaxErrors::default()`

File: snafu-derive/src/lib.rs (hunk 2 above)
Comment (shepmaster) [reply snafu-derive/src/lib.rs]: > eventually found that a couple custom types were simpler

I don't disagree, but would you mind expanding on bit on how you found the methods lacking?

File: compatibility-tests/compile-fail/tests/ui/attribute-duplication.rs (hunk 1 above)
Comment (shepmaster) [reply compatibility-tests/compile-fail/tests/ui/attribute-duplication.rs]: > I didn't get struct duplication checks working

Can you expand a bit on what problems arose?

Comment (tjkirch) [reply compatibility-tests/compile-fail/tests/ui/attribute-duplication.rs]: It was just that the enum-handling code's structure made it obvious what to do, and the struct-handling code wasn't as obvious for me, and I wasn't comfortable enough yet with the code style.

Now that I look back at it, I think there may only be one line to change, where we match `Source::From(t, e)` -- if we already have `Some`, then instead of setting `transformation`, we'd add an error.  Does that seem right?

File: snafu-derive/src/lib.rs (hunk 2 above)
Comment (tjkirch) [reply snafu-derive/src/lib.rs]: Sure - in particular it was `SyntaxErrors.only_valid_on`, which became the `OnlyValidOn` struct.  It had 3 arguments for descriptive strings, and 1 for the span.  The 3 string arguments were too easy for a developer to mix up, in my opinion -- you always had to check the docs.  With the struct, where you specify its fields at creation, it becomes obvious, and I think it reads very nicely.

(I think having those 3 fields are important to clarify the different types of errors, and clearly point out to the user what they've done wrong.)

File: snafu-derive/src/lib.rs (hunk 6 above)
Comment (tjkirch) [reply snafu-derive/src/lib.rs]: Sorry - sure, that's trivial.  I had just copied the `fn fmt` declaration from my current Rust source :)

File: snafu-derive/src/lib.rs (hunk 7 above)
Comment (tjkirch) [reply snafu-derive/src/lib.rs]: I'm not sure why I usually forget that `Default` is a thing, but this is an easy improvement I'll make. :)

File: compatibility-tests/compile-fail/tests/ui/attribute-duplication.rs (hunk 1 above)
Comment (tjkirch) [reply compatibility-tests/compile-fail/tests/ui/attribute-duplication.rs]: (And if that's the case, the same thing applies to enums; it's another type of duplication I hadn't caught.)

File: snafu-derive/src/lib.rs (hunk 2 above)
Comment (shepmaster) [reply snafu-derive/src/lib.rs]: > The 3 string arguments were too easy for a developer to mix up, in my opinion

I had a hunch this was it. For future reference, one thing I've seen Niko Matsakis do is pseudo-named-arguments:

```rust
struct Attribute(&'static str);
struct ValidOn(&'static str);
struct NotOn(&'static str);

fn only_valid_on(attribute: Attribute, valid_on: ValidOn, not_on: NotOn) {
    println!(
        "'{}' attribute is only valid on {}; attempted to place on this {}:",
        attribute.0, valid_on.0, not_on.0
    )
}

fn main() {
    only_valid_on(Attribute("a"), ValidOn("b"), NotOn("c"));
}
```

File: compatibility-tests/compile-fail/tests/ui/attribute-duplication.rs (hunk 1 above)
Comment (shepmaster) [reply compatibility-tests/compile-fail/tests/ui/attribute-duplication.rs]: > if we already have `Some`, then instead of setting `transformation`, we'd add an error

That would be fine, but we have [similar cases elsewhere](https://github.com/shepmaster/snafu/blob/da2d4aced258bdc443c90d8b3de6b127912e2af0/snafu-derive/src/lib.rs#L415), where we "pop one and exactly one" value from a `Vec`. 

Could we change these cases to push to a `Vec` and then reuse / repurpose that logic:

```rust
fn only_one<T>(v: Vec<T>) -> Option<T> { ... }

let from = only_one(froms).ok_or_else(...)
```

Comment (shepmaster) [reply compatibility-tests/compile-fail/tests/ui/attribute-duplication.rs]: Not 100% sure how to combine that with cases of zero-or-one, or as a not-immediately-fatal error.

Comment (tjkirch) [reply compatibility-tests/compile-fail/tests/ui/attribute-duplication.rs]: I'll give this a shot.

File: snafu-derive/src/lib.rs (hunk 2 above)
Comment (tjkirch) [reply snafu-derive/src/lib.rs]: I'm happy to go with whichever style you prefer :)

File: compatibility-tests/compile-fail/tests/ui/attribute-duplication.rs (hunk 1 above)
Comment (tjkirch) [reply compatibility-tests/compile-fail/tests/ui/attribute-duplication.rs]: I came up with something I'm pretty happy with - `AtMostOne`, which replaces those `Vec`s and creates errors for attributes beyond the first.  It handles the struct case and cleans up the enum cases, removes the weird indexing reverse, and builds nicely on `SyntaxErrors`.  I hope you like it.

It didn't replace `one_field_error` which wasn't as good a fit, but if you have ideas for merging the two I could do that.  (I imagine it'd take a `Fn` for the error generation, or something similar, but it seemed like more overhead than it was worth to replace this one case, when the simpler thing replaced all the rest.)

File: snafu-derive/src/lib.rs [hunk 8]
```diff
@@ -347,12 +627,15 @@ fn parse_snafu_struct(
...
-        Some((ty, expr)) => Transformation::Transform { ty, expr },
-        None => Transformation::None {
+    let (maybe_transformation, errs) = transformations.finish();
+    let transformation = maybe_transformation
+        .map(|(ty, expr)| Transformation::Transform { ty, expr })
+        .unwrap_or(Transformation::None {
             ty: inner.into_value().ty,
-        },
-    };
+        });
```
Comment (shepmaster) [inline snafu-derive/src/lib.rs]: FYI, [`unwrap_or_else`](https://doc.rust-lang.org/std/option/enum.Option.html#method.unwrap_or_else) and [`map_or_else`](https://doc.rust-lang.org/std/option/enum.Option.html#method.map_or_else) exist.

This currently unconditionally calls `into_value`, which is a *slight* inefficiency. I'm not really worried about it here or in the earlier case of it.

**Comment by tjkirch** [comment]
> This all seems great; thanks so much for putting up with all my pedantry!

Your knowledge and guidance are invaluable to the community.  I'm thankful to have been able to take advantage, while helping a project I love.  It inspires me to find other ways to contribute.  So, thank you!

> Is there anything else you think we need to add or discuss before merging?

Just the `or_else` fix that I'll make shortly.  I agree with you that the spans the new errors point to aren't ideal, but it would require a bit more wide-reaching changes that aren't strictly related, so I think it should be a separate issue.  I do think this change sets it up for success with how it accepts/stores spans, but the right ones need to be made available.  Do you agree that should be another issue?  I can make it if so.

File: snafu-derive/src/lib.rs (hunk 8 above)
Comment (tjkirch) [reply snafu-derive/src/lib.rs]: Oops; that line was a copy/paste, and I missed the change in behavior in this new context.  Will fix shortly.

File: compatibility-tests/compile-fail/tests/ui/attribute-duplication.stderr [hunk 9]
```diff
@@ -0,0 +1,71 @@
+error: Only one field can be designated as 'visibility'; found a duplicate here:
```
Comment (shepmaster) [inline compatibility-tests/compile-fail/tests/ui/attribute-duplication.stderr]: This isn't a field, is it? Shouldn't this say something like "multiple visibility attributes are not supported on an error enum"?

File: compatibility-tests/compile-fail/tests/ui/attribute-duplication.stderr [hunk 10]
```diff
@@ -0,0 +1,71 @@
...
+ --> $DIR/attribute-duplication.rs:7:10
+  |
+7 |     enum EnumError {
+  |          ^^^^^^^^^
+
+error: Only one field can be designated as 'source'; found a duplicate here:
```
Comment (shepmaster) [inline compatibility-tests/compile-fail/tests/ui/attribute-duplication.stderr]: How about we add "per variant" to this class of error?

File: compatibility-tests/compile-fail/tests/ui/attribute-duplication.stderr [hunk 11]
```diff
@@ -0,0 +1,71 @@
...
+18 |             source2: String,
+   |             ^^^^^^^
+
+error: Only one field can be designated as 'backtrace(delegate)'; found a duplicate here:
+  --> $DIR/attribute-duplication.rs:21:13
+   |
+21 |             source3: String,
+   |             ^^^^^^^
+
+error: Cannot have 'backtrace' and 'backtrace(delegate)' fields in the same enum
```
Comment (shepmaster) [inline compatibility-tests/compile-fail/tests/ui/attribute-duplication.stderr]: In the same enum or in the same variant?

File: compatibility-tests/compile-fail/tests/ui/attribute-duplication.stderr [hunk 12]
```diff
@@ -0,0 +1,71 @@
...
+24 |             backtrace1: String,
+   |             ^^^^^^^^^^
+
+error: Cannot have 'backtrace' and 'backtrace(delegate)' fields in the same enum
+  --> $DIR/attribute-duplication.rs:15:13
+   |
+15 |             source: Box<EnumError>,
+   |             ^^^^^^
+
+error: Only one field can be designated as 'display'; found a duplicate here:
```
Comment (shepmaster) [inline compatibility-tests/compile-fail/tests/ui/attribute-duplication.stderr]: ditto: not a field, but duplicate attrs per variant

File: compatibility-tests/compile-fail/tests/ui/attribute-duplication.stderr [hunk 13]
```diff
@@ -0,0 +1,71 @@
...
+  --> $DIR/attribute-duplication.rs:12:9
+   |
+12 |         AVariant {
+   |         ^^^^^^^^
+
+error: Only one field can be designated as 'visibility'; found a duplicate here:
```
Comment (shepmaster) [inline compatibility-tests/compile-fail/tests/ui/attribute-duplication.stderr]: ditto: not a field, but duplicate attrs per variant

File: compatibility-tests/compile-fail/tests/ui/attribute-duplication.stderr (hunk 9 above)
Comment (tjkirch) [reply compatibility-tests/compile-fail/tests/ui/attribute-duplication.stderr]: You're right, the context isn't as good as it should be, even without more detailed spans.  I'll work on improving these.

**Comment by tjkirch** [comment]
> I re-read some of the error text and it seems like a few errors got mixed up; am I losing it?

Nope!  I didn't have the right context in a few spots.  I've cleaned those up (easy to do by adding a context `location` to the `AtMostOne`) and tried to make the error wording more consistent with yours.

While doing this, I found a couple more cases of duplication I'm not catching yet, which I think I can fix fairly quickly with the same tools:

* We catch `backtrace*` on multiple fields in a variant, but not multiple `backtrace*` on a single field
* We don't catch `source(from)` and `source(flag)` together on a variant field (only wrong if flag=true?)

**Comment by shepmaster** [comment]
Thanks for the last push! Everything looks good to me now; are there any outstanding issues (other than trying to find better things to point to) we should address before merging and releasing?

**Comment by tjkirch** [comment]
> Thanks for the last push! Everything looks good to me now; are there any outstanding issues (other than trying to find better things to point to) we should address before merging and releasing?

There are a couple more cases of duplication that I mentioned in https://github.com/shepmaster/snafu/pull/109#issuecomment-513607010 -- they're a little harder to fix than I initially expected, but if you're not worried about those, or you'd like them done in a separate change, then no blockers that I know of. :)

**Comment by shepmaster** [comment]
Oh, I assumed that they were part of your last force push. I'm in no hurry, I just don't want to forget about this PR (something I unfortunately have a tendency to do). Please feel free to poke me when you have a chance to cycle back to those last few. Feel free to ask questions if you think it's something I can help with!

**Comment by tjkirch** [comment]
> Feel free to ask questions if you think it's something I can help with!

I'd like to clarify whether some combinations of attributes should be errors, to be sure I'm not going to change behavior.

* I'm guessing that `source(false)` and `source(from(...))` on the same field would be wrong, but I'm not sure - maybe the user is effectively saying "this isn't the source *directly*, but you can make it *from* ...".  Error?
* Would you want to allow `source(true)` and `source(from(...))` on the same field?  Thinking charitably, the user could be double-stating their intention, but it feels more like a confusion about the `from`.  Error?
* I'm guessing `backtrace(false)` and `backtrace(delegate)` on the same field would be wrong, though it's less clear - they could be trying to say "*this* shouldn't backtrace, look inside this delegate".  Still feels like a confusion to me.  Error?
* How about `backtrace(true)` and `backtrace(delegate)`?  That seems even weirder to me.  Error?

Separate from those, I think there are a couple more cases to clean up regardless:
* a field is only intended to have a backtrace delegate if it's a source, but a backtrace delegate on a non-source field is quietly ignored - I think that should be an error.
* a field shouldn't be a non-delegate backtrace and a source, but we quietly ignore the backtrace attribute on a source - I think it should be an error.

As for why it's more challenging than anticipated, aside from the above, it's the interplay between `has_backtrace`, `is_backtrace`, and `is_source` in the field handling code.  I understand what they do now, I'm just trying to find a clean pattern for accomplishing the same checks.  I lean toward adding sources/backtraces to the higher-level AtMostOne as soon as we find them, and using the `is_`/`has_` fields for special error checks (mentioned above) after the attribute loop.  I haven't tried that yet (it'd be nice to confirm the above confusions first) but I'll update this issue if I have trouble.

**Comment by shepmaster** [comment]
> * `source(true)` and `source(from(...))` on the same field
> * `source(false)` and `source(from(...))` on the same field

I could go either way on these. `true` and `from` together don't bother me. I see `true` as meaning "this is a source that cannot be named `source`" and `from` as "here is how you transform from one to the other". It may be better if `from` implied `true`, which it doesn't seem like it does currently.

`false` and `source` do seem like a point of confusion.

> * a non-delegate backtrace and a source

This is interesting. I wonder if we should have made it such that `#[snafu(source, backtrace)]` implicitly meant `delegate`. Then we wouldn't have needed the surface-level API for `backtrace(delegate)` at all. That might also remove some possibilities that we need to prevent.

Do you think that this would be confusing to end users?

> I lean toward adding sources/backtraces to the higher-level AtMostOne as soon as we find them, and using the `is_`/`has_` fields for special error checks (mentioned above) after the attribute loop

Sounds reasonable. I think you'd use the `AtMostOne` before/during the loop, call `finish` (collecting errors), then perform a `match` on the resulting `Option`s to ensure that there's only a reasonable combination.

**Comment by tjkirch** [comment]
> `true` and `from` together don't bother me

That makes sense to me, I'll make sure this still works.

> It may be better if `from` implied `true`

That makes sense and it's easy to do, I'll add that.

> `false` and `source` do seem like a point of confusion.

Ok, I'll make this an error.

> I wonder if we should have made it such that `#[snafu(source, backtrace)]` implicitly meant `delegate`

That's a good idea.  Any `backtrace` on a `source` would be the equivalent of today's `delegate`, and on a non-`source` wouldn't change.  I also lean toward that simpler API.  As a new user, I think it does have a slight potential for confusion, but no more than `delegate` itself, to be honest :)

I think I could do this.  (It'd be a third commit, of course.)

**Comment by shepmaster** [comment]
Sure. If at any point you think that this PR is getting too big, I definitely don't mind merging and creating some new issues / PRs to iterate.

**Comment by tjkirch** [comment]
> If at any point you think that this PR is getting too big, I definitely don't mind merging and creating some new issues / PRs to iterate.

I think the `delegate` change could definitely be separate; my concern there was just that merging these breaking changes, without having the remaining breaking change ready, could impact your ability to cut a release.  If you wanted to or had to for other reasons, you might need yet another release for the breaking `delegate` change.

If you're not worried about that, then all that would remain for this PR is preventing `source(false)` and `source(from(...))` on the same field (as a form of misuse).

**Comment by shepmaster** [comment]
We can merge this but not immediately release anything; bundling up the breaking changes. If we need to release a patch version, I can always fork from before these commits.

**Comment by tjkirch** [comment]
Ok!  I was being perhaps overly cautious.

To recap, here's the remaining work that I think could fit in this PR:
* Error on `source(false)` and `source(from)` on one field  (still allowing `source(true)` and `source(from)`)
* Error on multiple `source` or `backtrace` on one field

These could be other PRs, and I'll make issues for them:
* Making `source(from)` imply `source(true)`
* Removing `backtrace(delegate)` in favor of `backtrace` on a `source` field
* Improving the accuracy of the spans in the new errors

**Comment by tjkirch** [comment]
I've addressed the remaining work for this PR:
* `source(false)` and `source(from)` on a single field now cause an IncompatibleAttributes error
* Multiple `source` or `backtrace` attributes on a single field now error through new AtMostOne trackers

compile-fail tests were added for both.  If you like the approach I took, I think this PR is ready for merging.

With the restructuring, making `source(from)` imply `source(true)` for #124 should be a one-line change in another PR.  It also exposes a couple more duplicates in existing compile-fail tests, because the tests had multiple fields with `source(from)` that weren't counted as sources and therefore not as duplicates :)

**Comment by tjkirch** [comment]
Hmm.  Rust 1.18 fails with a lifetime issue: https://github.com/shepmaster/snafu/pull/109/checks?check_run_id=176857626

I'm not sure offhand what changed between Rust 1.18 and 1.30 that allows this slice to be considered static: https://github.com/shepmaster/snafu/compare/6cb438122453631443eb361840d9499768a9d54a..27504266fd1b266e101b22de103881e84af0a01c#diff-89e1fad4d2637dae0904dce6aba08394R487

@shepmaster - assuming you're OK with this general approach, would you recommend `IncompatibleAttributes` store a `Vec`, and I use a `vec![]` during construction to resolve that lifetime issue, or is there a better way?

**Comment by tjkirch** [comment]
I just learned about `cd compatibility-tests/v1_18 && rustup run 1.18.0 cargo test` so hopefully I can prevent bad pushes in the future - sorry about that.  Another tip I'm thinking about adding to a contributor's guide...

**Comment by tjkirch** [comment]
> I'm not sure offhand what changed between Rust 1.18 and 1.30 that allows this slice to be considered static: https://github.com/shepmaster/snafu/compare/6cb438122453631443eb361840d9499768a9d54a..27504266fd1b266e101b22de103881e84af0a01c#diff-89e1fad4d2637dae0904dce6aba08394R487

I think it might be this change: https://github.com/rust-lang/rust/issues/38865
...but there was also this one that I don't really understand :)  https://github.com/rust-lang/rust/pull/47408

**Comment by shepmaster** [comment]
> making `source(from)` imply `source(true)` for #124 should be a one-line change in another PR.

Nice!

> multiple fields with `source(from)` that weren't counted as sources and therefore not as duplicates

Makes sense 😉 

> what changed between Rust 1.18 and 1.30 that allows this slice to be considered static

Yeah, it'd be the automatic promotion of literals to statics, as you linked.

> or is there a better way

You can do the promotion by hand:

```rust
static FOO: &[&str] = &["source(false)", "source(from)"];
```
```rust
attributes: FOO,
```

> cd compatibility-tests/v1_18 && rustup run 1.18.0 cargo test

There's actually a "rust-toolchain" file in that directory, so you can just do `cd compatibility-tests/v1_18 && cargo test`. If that file wasn't there, I'd recommend doing `cd compatibility-tests/v1_18 && cargo +1.18.0 test` for fewer keystrokes.

> adding to a contributor's guide

That would probably be a good thing!

**Comment by shepmaster** [comment]
> bad pushes in the future - sorry about that

Not at all; that's the point of CI - to catch the things that we forget. That's why I try to be pedantic about adding new tests so frequently. I don't want to have to think about exactly what to remember!

It is useful to check in 1.18 locally only because it's comparatively so different to current Rust that a lot of little things sneak in while developing. My biggest mistake is around match ergonomics.

**Comment by tjkirch** [comment]
> You can do the promotion by hand:

Thanks, that worked!  I just pushed that fix, and hopefully this is everything we need to get this first PR merged.

> There's actually a "rust-toolchain" file in that directory, so you can just do cd compatibility-tests/v1_18 && cargo test. If that file wasn't there, I'd recommend doing cd compatibility-tests/v1_18 && cargo +1.18.0 test for fewer keystrokes.

That's good to know!  The other piece that was missing for me is understanding that `.cirrus.yml` lists some other steps that are required for each of the Rust versions - for example installing precise versions of dependencies that work with that version of Rust.  Without doing anything special, I was hitting errors building dependencies, like `backtrace` using the `crate` keyword in `use` statements.

Do you know if there's a way to run those locally without a lot of copy+paste?  I couldn't find any resources for replicating Cirrus CI locally.  I see they have an issue for making a CLI (https://github.com/cirruslabs/cirrus-ci-docs/issues/108) but didn't see anything else.

**Comment by shepmaster** [comment]
> a way to run those locally without a lot of copy+paste

Nothing I'm aware of. However, it's been pretty standard practice in my other crates to actually have most of the CI-related junk as separate shell scripts so that they are at least easier to access outside of CI (and you get syntax highlighting, etc.)

In this case, I wonder if we should just move those to the respective Cargo.toml files...

**Comment by shepmaster** [comment]
image

What an amazing first contribution! Thanks so much!

**Comment by tjkirch** [comment]
> it's been pretty standard practice in my other crates to actually have most of the CI-related junk as separate shell scripts
>
> In this case, I wonder if we should just move those to the respective Cargo.toml files...

Either of these sound like they would be pretty helpful.  I can create an issue if you like?

> What an amazing first contribution! Thanks so much!

Thanks again for the library, and for your great feedback here!  I hope to keep contributing.

**Comment by tjkirch** [comment]
> Either of these sound like they would be pretty helpful. I can create an issue if you like?

Ah, you already did #128!  Wow!
